flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/PreciseTimestamp.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.dataformat;
+
+import org.apache.flink.table.types.logical.TimestampType;
+import org.apache.flink.util.Preconditions;
+
+import java.sql.Timestamp;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.Arrays;
+
+import static org.apache.calcite.avatica.util.DateTimeUtils.timestampStringToUnixDate;
+import static org.apache.calcite.avatica.util.DateTimeUtils.unixTimestamp;
+import static org.apache.calcite.avatica.util.DateTimeUtils.unixTimestampToString;
+
+/**
+ * Immutable SQL TIMESTAMP type with nanosecond precision.
+ */
+public class PreciseTimestamp implements Comparable<PreciseTimestamp> {
+
+	public static final int MAX_PRECISION = TimestampType.MAX_PRECISION;
+
+	public static final int MIN_PRECISION = TimestampType.MIN_PRECISION;
+
+	private final long milliseconds;
+
+	private final int nanoseconds;
+
+	private final int precision;
+
+	private PreciseTimestamp(long milliseconds, int nanoseconds, int precision) {
+		Preconditions.checkArgument(precision >= MIN_PRECISION && precision <= MAX_PRECISION);
+		Preconditions.checkArgument(nanoseconds >= 0 && nanoseconds <= 999_999_999);
```

Comment:
So `milliseconds` contains mills, and `nanoseconds` contains mills too?
-1 for the design,  `milliseconds` and `nanoseconds` should not have overlap.

Reply:
Performance reason to retain milli part in **milliseconds**:
We always need the milli part in some operations, we can perform "milliseconds without milli part + nanoseconds / 1000_000" once and use it everywhere.

Reply:
So I think you can just remove mills in nanoseconds, because it is totally useless.
I think we can change name to nanoOfMills, just contain micros and nanos.